tests/unit_tests/cpu/covariance_transport.cpp

```diff
@@ -13,112 +13,222 @@
+
 // google-test include(s).
 #include <gtest/gtest.h>
 
 using namespace detray;
 
+// Algebra types
 using matrix_operator = standard_matrix_operator<scalar>;
 using transform3 = test::transform3;
 using vector3 = typename transform3::vector3;
 using intersection_t = intersection2D<surface<>, transform3>;
 
-// Setups
-constexpr const scalar tolerance = scalar(2e-2);
-const vector3 z_axis{0.f, 0.f, 1.f};
-const vector3 B{0.f, 0.f, 1.f * unit<scalar>::T};
+// Mask types to be tested
+using rectangle_type = detray::mask<detray::rectangle2D<>>;
+using trapezoid_type = detray::mask<detray::trapezoid2D<>>;
+using ring_type = detray::mask<detray::ring2D<>>;
+using cylinder_type = detray::mask<detray::cylinder2D<>>;
+using straw_wire_type = detray::mask<detray::line<false>>;
+using cell_wire_type = detray::mask<detray::line<true>>;
+
+// Test class for covariance transport
+template <typename T>
+class HelixCovarianceTransportValidation : public ::testing::Test {
+    public:
+    // Environment Setup
+    const vector3 B{0.f, 0.f, 1.f * unit<scalar>::T};
+    const vector3 y_axis{0.f, 1.f, 0.f};
+    const vector3 z_axis{0.f, 0.f, 1.f};
+    static constexpr const scalar mask_tolerance{1e-3f};
+    static constexpr const scalar tolerance{3e-2f};
+
+    // Test types
+    using mask_type = typename T::first_type;
+    using local_frame_type = typename mask_type::local_frame_type;
+    using helix_intersector_type = typename T::second_type;
+
+    // First mask at the origin is always rectangle
+    using first_mask_type = rectangle_type;
+    using first_local_frame_type = typename first_mask_type::local_frame_type;
+    using first_helix_intersector_type =
+        detail::helix_plane_intersector<intersection_t>;
+
+    // Transform3 type
+    using transform3_type = typename local_frame_type::transform3_type;
+    // Scalar type
+    using scalar_type = typename transform3_type::scalar_type;
+
+    // Vector and matrix types
+    using free_vector = typename local_frame_type::free_vector;
+    using bound_vector = typename local_frame_type::bound_vector;
+    using bound_matrix = typename local_frame_type::bound_matrix;
+    using free_to_bound_matrix =
+        typename local_frame_type::free_to_bound_matrix;
+    using bound_to_free_matrix =
+        typename local_frame_type::bound_to_free_matrix;
+    using free_matrix = typename local_frame_type::free_matrix;
+
+    // Create first mask
+    first_mask_type create_first_mask() {
+        return rectangle_type{0u, 50 * unit<scalar>::mm, 50 * unit<scalar>::mm};
+    }
 
-// Algebra objects
-constexpr const cartesian2<transform3> c2;
-constexpr const detail::helix_plane_intersector<intersection_t> hpi;
+    // Create mask for test types
+    // 1. rectangle
+    template <typename M = mask_type,
+              typename std::enable_if_t<std::is_same_v<M, rectangle_type>,
+                                        bool> = true>
+    mask_type create_mask() {
```

Comment:
This would probably be less verbose with function template specialization. However, I guess it would be easiest with just a tuple of masks and then ```detail::get<mask_type>(maks_tuple)```.

Reply:
Done